### PR TITLE
feat: Migrate from docker-compose to kubernetes

### PR DESCRIPTION
**Pull request by dwijesh-shah-crest**
Pytest-splunk-addon migration from docker to kubernetes.

- All the tests are executing by spinning up the kubernetes resources (Splunk, SC4S, UF). (Github Actions, local)
- Once all the tests will be executed, in the SessionFinish all the kubernetes resources will be destroyed. (If --keep-alive is not passed)
- To execute addons test using --splunk-type=kubernetes, please find the steps here: https://github.com/splunk/pytest-splunk-addon/blob/test/migrate-k8s-poc/README-test.md

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -94,56 +93,29 @@ Create a test file in the tests folder
...
         --splunk-port=8089 \
         --splunk-password=Changed@11 \
         -v
 
-Run pytest with the add-on and SA-eventgen installed and enabled in docker
+Run pytest with the add-on and SA-eventgen installed and enabled in kubernetes
+
+For UCC based addons generate addon SPL with `ucc-gen --ta-version=<package/default/app.conf/id.version>`_.
+
+For Non-UCC based addons generate addon SPL of format ``<package/default/app.conf/id.name>-<package/default/app.conf/id.version>.spl``
```
Comment (arys-splunk) [inline README.rst:111]: Does `ucc-gen` generate different name for UCC-based and non-UCC-based add-ons?

File: .github/workflows/build-test-release.yml [hunk 2]
```diff
@@ -123,37 +123,72 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        splunk-version: ["8.1","8.2"]
```
Comment (arys-splunk) [inline .github/workflows/build-test-release.yml:126]: What is this change for?

File: .github/workflows/build-test-release.yml [hunk 3]
```diff
@@ -175,21 +210,36 @@ jobs:
...
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
-        splunk-version: [8.1, 8.2]
+        test-marker: ['test_report', 'splunk_connection_kubernetes', 'splunk_app_fiction', 'splunk_app_broken','splunk_app_cim_fiction','splunk_app_cim_broken','splunk_fiction_indextime','splunk_fiction_indextime_broken','splunk_setup_fixture','splunk_app_requirements','splunk_app_requirements_modinput','splunk_app_requirements_uf','splunk_app_requirements_scripted']
+        splunk-version: ['8.1','8.2']
+        os: [ubuntu-latest]
```
Comment (arys-splunk) [inline .github/workflows/build-test-release.yml:224]: I think it was intended to run on all OS-es, but the configuration was incorrect, are we only supporting Ubuntu now?

File: docs/sample_generator.rst [hunk 4]
```diff
@@ -59,7 +59,7 @@ input_type = modinput | scripted_input | syslog_tcp | file_monitor | windows_inp
     * For example, in an Add-on, a sourcetype "alert" is ingested through syslog in live environment, provide input_type=syslog_tcp.
     
     .. warning::
-        uf_file_monitor input_type will only work with splunk-type=docker.
+        uf_file_monitor input_type will only work with splunk-type=kubernetesss.
```
Comment (arys-splunk) [inline docs/sample_generator.rst:62]: Typo here.

File: pyproject.toml [hunk 5]
```diff
@@ -42,16 +42,12 @@ faker = ">=4.1,<12.0"
 pytest-xdist = "*"
 filelock = "^3.0"
 pytest-ordering = "*"
-lovely-pytest-docker = { version="^0", optional = true }
 junitparser = "^2.2.0"
 addonfactory-splunk-conf-parser-lib = "^0.3.3"
+crudini = "0.9.3"
```
Comment (arys-splunk) [inline pyproject.toml:47]: You do not need `cruduni` here, please checkout `addonfactory-splunk-conf-parser-lib`, it does pretty much the same thing you need.

**Review by arys-splunk — COMMENTED** [review]
What is the time difference in running these tests using Kubernetes comparing to Docke?

File: .github/workflows/build-test-release.yml (hunk 3 above)
Comment (arys-splunk) [reply .github/workflows/build-test-release.yml:224]: Actually, found an original code for this: https://github.com/splunk/pytest-splunk-addon/pull/426/files#diff-227729f71180f01a43dec20798948144523f45c9e4f23555469002e7178ce4b4R197-R205

**Comment by arys-splunk** [comment]
Let's not merge this one to main, but to develop.

File: README.rst [hunk 6]
```diff
@@ -94,56 +93,29 @@ Create a test file in the tests folder
...
-        - "8000"
-        - "8089"
-        environment:
-        - SPLUNK_PASSWORD=Changed@11
-        - SPLUNK_START_ARGS=--accept-license
-
 Run pytest with the add-on and SA-eventgen installed and enabled in an external Splunk deployment
 
-.. code::: bash
+.. code::: python3
```
Comment (uoboda-splunk) [inline README.rst:98]: It is not visible right now. Maybe `code::`

File: README.rst [hunk 7]
```diff
@@ -94,56 +93,29 @@ Create a test file in the tests folder
...
+Run pytest with the add-on and SA-eventgen installed and enabled in kubernetes
+
+Generate addon SPL with `ucc-gen`_.
+
+For third-party addons generate addon SPL.
 
-.. code::: bash
+Place the generated addon spl in tests/src/<addon>.spl
+
+.. code::: python3
```
Comment (uoboda-splunk) [inline README.rst:115]: Not visible too

File: README.rst [hunk 8]
```diff
@@ -94,56 +93,29 @@ Create a test file in the tests folder
...
-        --splunk-type=external \
         --splunk-host=splunk \
         --splunk-port=8089 \
         --splunk-password=Changed@11 \
         -v
 
-Run pytest with the add-on and SA-eventgen installed and enabled in docker
+Run pytest with the add-on and SA-eventgen installed and enabled in kubernetes
+
+Generate addon SPL with `ucc-gen`_.
```
Comment (uoboda-splunk) [inline README.rst:109]: Maybe info about splunk-operator deployment also here?

File: pytest_splunk_addon/splunk.py [hunk 9]
```diff
@@ -485,45 +487,123 @@ def sc4s(request):
...
+        else:
+            os.environ["TEST_RUNNER_DIRECTORY_UF"] = os.getenv("TEST_RUNNER_DIRECTORY")
+        if "\\" in os.getenv("TEST_RUNNER_DIRECTORY_UF"):
+            os.environ["TEST_RUNNER_DIRECTORY_UF"] = os.getenv(
+                "TEST_RUNNER_DIRECTORY_UF"
+            ).replace("\\", "/")
+        update_k8s_manifest_files(folder=current_path, file="uf_deployment")
+        LOGGER.info("Setting up UF")
+        kubernetes_helper_uf = KubernetesHelper()
+        kubernetes_helper_uf.create_k8s_resource_from_yaml(
```
Comment (uoboda-splunk) [inline pytest_splunk_addon/splunk.py:554]: Is it possible to create resources in parallel to make it faster?

File: README.rst [hunk 10]
```diff
@@ -94,56 +93,29 @@ Create a test file in the tests folder
...
+
+For third-party addons generate addon SPL.
 
-.. code::: bash
+Place the generated addon spl in tests/src/<addon>.spl
+
+.. code::: python3
 
         pytest \
-        --splunk-password=Changed@11 \
```
Comment (uoboda-splunk) [inline README.rst:146]: Should param `--splunk-password` setup password for splunk? If yes I think it is not working

File: pytest_splunk_addon/splunk.py [hunk 11]
```diff
@@ -541,52 +621,234 @@ def uf_external(request):
...
+            "{0}".format(os.getenv("TEST_RUNNER_DIRECTORY")), "tests", "src"
+        )
+        splunk_spl_name = [
+            filename
+            for filename in os.listdir(spl_path)
+            if filename.startswith(("{0}-").format(splunk_addon_name))
+        ]
+        SPLUNK_ADDON = str(splunk_spl_name[0])
+        LOGGER.info(SPLUNK_ADDON)
+    except Exception as e:
```
Comment (uoboda-splunk) [inline pytest_splunk_addon/splunk.py:706]: Can we also check here version like `1.0.0` because it will also generate errors in tests right?

**Comment by dwijesh-shah-crest** [comment]
> What is the time difference in running these tests using Kubernetes comparing to Docke?

By implementing parallelisation, the PSA CI/CD test execution time has been reduced to approximately 35 minutes.

File: README.rst (hunk 1 above)
Comment (dwijesh-shah-crest) [reply README.rst]: Updated the README.

File: .github/workflows/build-test-release.yml (hunk 3 above)
Comment (dwijesh-shah-crest) [reply .github/workflows/build-test-release.yml]: Previously it was only executing on ubuntu-latest even though mac-os and windows has been passed. A JIRA has been created to implement the functionality to test on windows and mac in CI/CD.
Ref: https://splunk.atlassian.net/browse/AQA-69

File: pytest_splunk_addon/splunk.py (hunk 9 above)
Comment (dwijesh-shah-crest) [reply pytest_splunk_addon/splunk.py:554]: SC4S depends on Splunk setup, I think it won't be possible for now.

File: README.rst (hunk 10 above)
Comment (dwijesh-shah-crest) [reply README.rst:146]: For --splunk-type=kubernetes password and hec-token will be generated dynamically which will be used while executing tests.

File: pytest_splunk_addon/splunk.py (hunk 11 above)
Comment (dwijesh-shah-crest) [reply pytest_splunk_addon/splunk.py:706]: Addon SPL will be of format `Addon_Name`-`version` ; version can be anything, we only require SPL with the given format, which will be installed while spinning up the splunk standalone instance.

**Comment by dwijesh-shah-crest** [comment]
> 



> The proposed solution doesn't work on mac and windows. Also, we should consider and redesign the architecture IMHO we cannot base on subprocess and execute the external command as long PSA is a python package. Things like jq, kubectl and crudini should be replaced by Python implementation. eg kubectl implementation should be done using official https://github.com/kubernetes-client/python lib. We should also avoid hardcoded sleep with magic numbers and improve logging. Spinning up tests can hang for a long time (1h+) without any information. Any manual precondition should be automated. It will be good if we will make a health check before spinning up env.

Crudini has been replaced with addonfactory-splunk-conf-parser-lib, and jq is also replaced with re. Kubectl utility has been replaced with kubernetes-client and shell scripts has been replaced with the python based methods to create and destroy kubernetes resources. Also, retry mechanism has been implemented for kubernetes resources creation.

**Comment by dwijesh-shah-crest** [comment]
> Let's not merge this one to main, but to develop.

PR for develop: https://github.com/splunk/pytest-splunk-addon/pull/569, will close this after PR of develop gets merged.

**Comment by dwijesh-shah-crest** [comment]
PR for main: https://github.com/splunk/pytest-splunk-addon/pull/575, will close this after mentioned PR will get merged.

**Pull request closed without merging**